oss/ossQueue.hpp

```diff
@@ -0,0 +1,127 @@
+
+#include "core.hpp"
+
+namespace emeralddb {
+namespace oss {
+    template<typename Data>
+    class ossQueue {
+    private:
+        std::queue<Data> _queue; 
+        boost::mutex _mutex; 
+        boost::condition_variable _cond;
+    public:
+
+        // method to get the length of current queue  
+        // everytime we try to get the length of the queue, we need to add a exclusive lock upon the get length operation 
+        // which means it is a critical section should be protected by exclusive lock to prevent other threads update the queue's length or queue's elements  
+        unsigned int size() {
+            boost::mutex::scoped_lock lock(_mutex); 
+            return (unsigned int) _queue.size(); 
+        }// size 
+ 
+        // method to push one item to the queue  
+        void push(Data const& data) {
+            // here we create a lock based on the class global variable the _mutex
+            boost::mutex::scoped_lock lock(_mutex);
+       
+            // here if the thread succeeed in racing the lock it will come here to operate pushing data to the queue 
+            _queue.push(data);
+
+            // then after push data finish, current thread will execute release the lock 
+            lock.unlock(); 
+ 
+            // then the thread will execute the notify_one to invoke one of the multiple blocked threads to retrieve the lock again 
+            // let the condition to notify one of the multiple threads that are waiting/blocking on the lock 
+            _cond.notify_one(); 
+        }  // push 
+       
+        // method to checkout whether the queue is empty 
+        bool empty() const {
+            // here we call lock method lock on _mutex
+            // when function executed finished, the lock will release the _mutex automatically 
+            boost::mutex::scoped_lock lock(_mutex); 
+            return _queue.empty(); 
+        } 
+      
+        // method that attempts to pop one element from the front of the queue 
+        // if pop success it will be referred by the parameter the Data & and return value will be set to true 
+        // else the passing reference variable will be set to null and the reutrn value will be set to false 
+        bool try_pop(Data &value ) {
+            boost::mutex::scoped_lock lock(_mutex); 
+            if (_queue.empty()) {
+                return false;  
+            }
+            value = _queue.front(); 
+            _queue.pop(); 
+            return true;  
+       }
+
+       // method to retrieve one element from the queue , if the queue is empty just wait without timeout 
+       // until the queue has one element retrieve the element and finish 
+       void wait_and_pop(Data &value) {
+           boost::mutex::scoped_lock lock(_mutex); 
+           // loop to wait and detect wehther the queue is not empty 
+           while (_queue.empty()) {
+               // every time the condition wait on the lock it will release the lock 
```

Comment:
every time the thread executes the condition#wait function the thread's lock will be released and handed over to other threads